### PR TITLE
Add support to CDC-MBIM USB host driver

### PR DESCRIPTION
**Pull request by acassis**
## Summary
Add support to CDC-MBIM USB host driver

This driver was created by Adam Porter and posted on NuttX
mailing list at Google Group on Nov 14 2019
## Impact
USB and NET modifications
## Testing
In progress

### CODE REVIEW

File: net/netdev/netdev_register.c [hunk 1]
```diff
@@ -313,6 +314,11 @@ int netdev_register(FAR struct net_driver_s *dev, enum net_lltype_e lltype)
             devfmt   = NETDEV_TUN_FORMAT;
             break;
 #endif
+        case NET_LL_MBIM:
+          dev->d_llhdrlen = 0;
+          dev->d_pktsize = 1200;
+          devfmt = NETDEV_MBIM_FORMAT;
+          break;
```
Comment (patacongo) [inline net/netdev/netdev_register.c:321]: Appears to be an alignment problem with default.

File: net/netdev/netdev_register.c [hunk 2]
```diff
@@ -55,6 +55,7 @@
 #define NETDEV_PAN_FORMAT   "pan%d"
 #define NETDEV_WLAN_FORMAT  "wlan%d"
 #define NETDEV_WPAN_FORMAT  "wpan%d"
+#define NETDEV_MBIM_FORMAT  "wwan%d"
```
Comment (patacongo) [inline net/netdev/netdev_register.c:58]: Why is MBIN treated separately.  Shouldn't the be NETDEV_WWLAN_FORMAT?  There are other wwlan as well, right?

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 3]
```diff
@@ -0,0 +1,2578 @@
...
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:35]: No new files may come into the repository with BSD headers.  The cannot be merge until the is changed to an Apache header.

File: net/netdev/netdev_register.c (hunk 1 above)
Comment (acassis) [reply net/netdev/netdev_register.c:321]: Good finding Greg! I will fix it!

File: net/netdev/netdev_register.c (hunk 2 above)
Comment (acassis) [reply net/netdev/netdev_register.c:58]: I think he tried to keep the consistency, wlan could conflict case a board has WiFi and the USB GSM Modem.

File: drivers/usbhost/usbhost_cdcmbim.c (hunk 3 above)
Comment (acassis) [reply drivers/usbhost/usbhost_cdcmbim.c:35]: Ok, I will fix it too. Thank you for reviewing it!

File: include/nuttx/net/net.h [hunk 4]
```diff
@@ -155,7 +155,8 @@ enum net_lltype_e
   NET_LL_BLUETOOTH,    /* Bluetooth */
   NET_LL_IEEE80211,    /* IEEE 802.11 */
   NET_LL_IEEE802154,   /* IEEE 802.15.4 MAC */
-  NET_LL_PKTRADIO      /* Non-standard packet radio */
+  NET_LL_PKTRADIO,     /* Non-standard packet radio */
+  NET_LL_MBIM
```
Comment (patacongo) [inline include/nuttx/net/net.h:159]: I don't think this is correct.

**Comment by btashton** [comment]
You cannot just slap the Apache license on this without getting his approval...

**Comment by acassis** [comment]
> You cannot just slap the Apache license on this without getting his approval...

Since the original code is BSD I think it is possible, but I should have kept the BSD license stand below the Apache license.

File: net/netdev/netdev_register.c (hunk 2 above)
Comment (patacongo) [reply net/netdev/netdev_register.c:58]: No, the corrent naming here must be  NETDEV_WWAN_FORMAT.

**Comment by patacongo** [comment]
> 
> 
> > You cannot just slap the Apache license on this without getting his approval...
> 
> Since the original code is BSD I think it is possible, but I should have kept the BSD license stand below the Apache license.

No, no dual licnese.  No, it is not possible to change the license header without consent from Adam.

**Comment by acassis** [comment]
> > > You cannot just slap the Apache license on this without getting his approval...
> > 
> > 
> > Since the original code is BSD I think it is possible, but I should have kept the BSD license stand below the Apache license.
> 
> No, no dual licnese. No, it is not possible to change the license header without consent from Adam.

Yes, I'm contacting him to discuss about it.

File: net/netdev/netdev_register.c [hunk 5]
```diff
@@ -313,6 +314,11 @@ int netdev_register(FAR struct net_driver_s *dev, enum net_lltype_e lltype)
             devfmt   = NETDEV_TUN_FORMAT;
             break;
 #endif
+        case NET_LL_MBIM:
+          llhdrlen = 0;
+          pktsize  = 1200;
+          devfmt   = NETDEV_MBIM_FORMAT;
+          break;
```
Comment (patacongo) [inline net/netdev/netdev_register.c:321]: That code sequence must be enclosed in conditional compilation base on CONFIG_USBHOST_CDCMBIM as are all other MAC layers.

File: include/nuttx/net/net.h (hunk 4 above)
Comment (patacongo) [reply include/nuttx/net/net.h:159]: What concerns me most is that NET_LL_MBIN is defined but used only in netdev_register().  For other MAC layers, there are numerous references throughout the networking (see below).  My gut says that the the NET_LL_MBIM must be integrated into the network in many of these places too.  Certainly ALL of these must be reviewed and those that require modification must be fixed:

    $ grep -rl NET_LL_ net
    net/arp/arp_send.c
    net/arp/arp_table.c
    net/bluetooth/bluetooth_finddev.c
    net/bluetooth/bluetooth_recvfrom.c
    net/bluetooth/bluetooth_sendto.c
    net/devif/devif_forward.c
    net/devif/devif_poll.c
    net/devif/ipv4_input.c
    net/devif/ipv6_input.c
    net/icmp/icmp_input.c
    net/icmp/icmp_poll.c
    net/icmp/icmp_recvfrom.c
    net/icmp/icmp_sendto.c
    net/icmpv6/icmpv6_advertise.c
    net/icmpv6/icmpv6_input.c
    net/icmpv6/icmpv6_poll.c
    net/icmpv6/icmpv6_radvertise.c
    net/icmpv6/icmpv6_recvfrom.c
    net/icmpv6/icmpv6_rsolicit.c
    net/icmpv6/icmpv6_sendto.c
    net/icmpv6/icmpv6_solicit.c
    net/ieee802154/ieee802154_finddev.c
    net/ieee802154/ieee802154_recvfrom.c
    net/ieee802154/ieee802154_sendto.c
    net/igmp/igmp_input.c
    net/igmp/igmp_poll.c
    net/igmp/igmp_send.c
    net/ipforward/ipfwd_forward.c
    net/ipforward/ipfwd_poll.c
    net/ipforward/ipv4_forward.c
    net/ipforward/ipv6_forward.c
    net/mld/mld_poll.c
    net/mld/mld_query.c
    net/mld/mld_send.c
    net/neighbor/neighbor_ethernet_out.c
    net/neighbor/neighbor_out.c
    net/netdev/netdev_default.c
    net/netdev/netdev_ioctl.c
    net/netdev/netdev_lladdrsize.c
    net/netdev/netdev_register.c
    net/pkt/pkt_poll.c
    net/pkt/pkt_sockif.c
    net/procfs/netdev_statistics.c
    net/sixlowpan/sixlowpan_framelist.c
    net/sixlowpan/sixlowpan_send.c
    net/sixlowpan/sixlowpan_tcpsend.c
    net/sixlowpan/sixlowpan_udpsend.c
    net/sixlowpan/sixlowpan_utils.c
    net/tcp/tcp_conn.c
    net/tcp/tcp_input.c
    net/tcp/tcp_ipselect.c
    net/tcp/tcp_recvfrom.c
    net/tcp/tcp_recvwindow.c
    net/tcp/tcp_send.c
    net/tcp/tcp_sendfile.c
    net/tcp/tcp_send_buffered.c
    net/tcp/tcp_send_unbuffered.c
    net/udp/udp_callback.c
    net/udp/udp_conn.c
    net/udp/udp_input.c
    net/udp/udp_ipselect.c
    net/udp/udp_recvfrom.c
    net/udp/udp_send.c
    net/udp/udp_sendto_buffered.c
    net/utils/net_icmpchksum.c
    net/utils/net_ipchksum.c

Comment (patacongo) [reply include/nuttx/net/net.h:159]: Certainly some of those hits are not relevant (liked sixlowpan).  But others point out significant missing logic that must be fixed.

File: net/netdev/netdev_register.c (hunk 1 above)
Comment (patacongo) [reply net/netdev/netdev_register.c:321]: Alignment is still wrong.

File: include/nuttx/net/net.h (hunk 4 above)
Comment (patacongo) [reply include/nuttx/net/net.h:159]: The integration into net/procfs/netdev_statistics.c is essential or else the ifconfig command will not work.  net/netdev/netdev_lladdrsize.c is used to report the size of the MAC header.

I don't think any of the forwarding applies.  That is only when we have to reformat packets in order to forward them.  I would need to check, but I think that the IOCTL stuff is only necessary for MAC-specific IOCTL support.

I would have to examine the remaining to see what is and is not relevant.

Comment (patacongo) [reply include/nuttx/net/net.h:159]: Related changes are also required in apps/ for selecting the right protocols and device names.  Perhaps other things.  apps/netutils/netlib will need to be checked.

File: include/nuttx/usb/cdc.h [hunk 6]
```diff
@@ -890,15 +930,15 @@ struct cdc_linecoding_s
 struct cdc_linestatus_s
 {
   uint8_t size[2];   /* wLength, Size of this structure, in bytes */
-  uint8_t ringer[4]; /* dwRingerBitmap, Ringer Configuration bitmap for this line */
+  uint8_t ringer[4]; /* dwRingerBitmap, Ringer Conf bitmap for this line */
```
Comment (patacongo) [inline include/nuttx/usb/cdc.h:933]:
```suggestion
  uint8_t ringer[4]; /* dwRingerBitmap, Ringer Configuration bitmap for this
                              * line */
```

Conf is not an acceptable abbreviation for anything.

File: include/nuttx/net/net.h (hunk 4 above)
Comment (patacongo) [inline include/nuttx/net/net.h:159]:
```suggestion
  NET_LL_MBIM        /* CDC-MBIM USB host driver */
```

File: net/netdev/netdev_register.c (hunk 5 above)
Comment (patacongo) [inline net/netdev/netdev_register.c:321]:
```suggestion
          break;
```

Comment (patacongo) [inline net/netdev/netdev_register.c:321]:
```suggestion
#ifdef CONFIG_USBHOST_CDCMBIM
          case NET_LL_MBIM:
            llhdrlen = 0;
            pktsize  = 1200;
            devfmt   = NETDEV_MBIM_FORMAT;
            break;
#endif
```

**Comment by aport** [comment]
Hi Alan, Greg, and team.

I'm okay with the license change.

A few things about this code:

1) It's old
2) It's incomplete
3) It's probably quite buggy
4) It needs a lot of work

It does actually work though, so it has that going for it.

The driver itself is mostly useless without a corresponding user app to send the MBIM commands. However writing this type of program is very simple and the MBIM specification is fully open.

I do not have access to any hardware that would allow me to test and/or continue developing this code. I hope that someone somewhere finds it useful.

File: net/netdev/netdev_register.c (hunk 2 above)
Comment (acassis) [reply net/netdev/netdev_register.c:58]: Agree!

**Comment by patacongo** [comment]
> 
> 
> Hi Alan, Greg, and team.
> 
> I'm okay with the license change.
> 
> A few things about this code:
> 
>     1. It's old
> 
>     2. It's incomplete
> 
>     3. It's probably quite buggy
> 
>     4. It needs a lot of work
> 
> 
> It does actually work though, so it has that going for it.
> 
> The driver itself is mostly useless without a corresponding user app to send the MBIM commands. However writing this type of program is very simple and the MBIM specification is fully open.
> 
> I do not have access to any hardware that would allow me to test and/or continue developing this code. I hope that someone somewhere finds it useful.

Thanks for the permission with the naming change.  Perhaps we could mark it experimental.  however, I do think that:

1. The implementation should be complete.  There are a few critical pieces missing, especially those that would permit ifconfig to work.
2. We need to have a configuration that we can use to at least verify that it still compiles.  Do you have a defconfig that we could use for testing.

This those things address (and the additional items from the comments), I think that we could provisionally merge the code.  I would approve the merge under those conditions.

File: drivers/usbhost/Kconfig [hunk 7]
```diff
@@ -262,6 +262,15 @@ config USBHOST_CDCACM_TXBUFSIZE
 
 endif # USBHOST_CDCACM
 
+config USBHOST_CDCMBIM
+	bool "CDC/MBIM support"
+	default n
+	depends on USBHOST_HAVE_ASYNCH && !USBHOST_BULK_DISABLE && !USBHOST_INT_DISABLE
+	select USBHOST_ASYNCH
+	---help---
```
Comment (patacongo) [inline drivers/usbhost/Kconfig:270]: Please add:

    depends on EXPERIMENTAL

File: net/netdev/netdev_register.c [hunk 8]
```diff
@@ -313,6 +314,11 @@ int netdev_register(FAR struct net_driver_s *dev, enum net_lltype_e lltype)
             devfmt   = NETDEV_TUN_FORMAT;
             break;
 #endif
+        case NET_LL_MBIM:
+          llhdrlen = 0;
+          pktsize  = 1200;
+          devfmt   = NETDEV_WWAN_FORMAT;
+          break;
```
Comment (patacongo) [inline net/netdev/netdev_register.c:321]:
```suggestion
#ifdef CONFIG_USBHOST_CDCMBIM
          case NET_LL_MBIM:
            llhdrlen = 0;
            pktsize  = 1200;
            devfmt   = NETDEV_WWAN_FORMAT;
            break;
#endif
```

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 9]
```diff
@@ -0,0 +1,2562 @@
...
+#include <nuttx/usb/usbhost.h>
+
+#define CDCMBIM_NETBUF_SIZE 8192
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* put in cdc.h */
+#define USB_CDC_SEND_ENCAPSULATED_COMMAND 0x00
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:59]: Blank space needed after comment

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 10]
```diff
@@ -0,0 +1,2562 @@
...
+  cdcwdm_write,  /* write */
+  NULL,          /* seek */
+  NULL,          /* ioctl */
+  cdcwdm_poll    /* poll */
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+  , NULL         /* unlink */
+#endif
+};
+
+/* This is a bitmap that is used to allocate device names /dev/cdc-wdm[n]. */
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:365]: Looks like a long line.  Is it?

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 11]
```diff
@@ -0,0 +1,2562 @@
...
+                           priv, 0);
+        }
+    }
+
+errout:
+  usbhost_givesem(&priv->exclsem);
+  return ret;
+}
+
+static ssize_t cdcwdm_write(FAR struct file *filep, FAR const char *buffer,
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:511]: Is that a long line?

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 12]
```diff
@@ -0,0 +1,2562 @@
...
+    }
+
+  /* Would be nice if CTRLIN would return how many bytes it read... */
+
+  struct mbim_header_s
+  {
+    uint8_t type[4];
+    uint8_t len[4];
+  };
+
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:941]: This is a C89 violation and must be fixed.  Move this structure to the top of the function.

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 13]
```diff
@@ -0,0 +1,2562 @@
...
+  struct mbim_header_s *hdr = (struct mbim_header_s *)priv->comm_rxbuf;
+  uint32_t len = usbhost_getle32(hdr->len);
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:943]: More C89 violations.  Same thing.

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 14]
```diff
@@ -0,0 +1,2562 @@
...
+ *   Standard character driver poll method.
+ *
+ ****************************************************************************/
+
+static int cdcwdm_poll(FAR struct file *filep, FAR struct pollfd *fds,
+                       bool setup)
+{
+  FAR struct inode           *inode;
+  FAR struct usbhost_cdcmbim_s *priv;
+  int                         ret = OK;
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:569]: Yech.  Pick one:  Aligned local variables, or ragged.  Either are find, but don't mix.

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 15]
```diff
@@ -0,0 +1,2562 @@
...
+  if (priv->bulkinbytes < (int16_t)(sizeof(struct usb_cdc_ncm_nth16_s) +
+                                    sizeof(struct usb_cdc_ncm_ndp16_s)))
+    {
+      goto out;
+    }
+
+  /* Parse the NTB header */
+
+  struct usb_cdc_ncm_nth16_s *nth =
+    (struct usb_cdc_ncm_nth16_s *)priv->rxnetbuf;
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:832]: C89 violation.  Move declaration of nth to the top of the function.

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 16]
```diff
@@ -0,0 +1,2562 @@
...
+      uerr("Invalid NTH signature\n");
+      goto out;
+    }
+
+  uint16_t block_len = usbhost_getle16(nth->block_length);
+
+  if (block_len > priv->bulkinbytes)
+    {
+      uerr("Block length larger than rx buffer\n");
+      goto out;
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:845]: C89 violation.  Move declaration of block_len to the top of the file.

File: drivers/usbhost/usbhost_cdcmbim.c [hunk 17]
```diff
@@ -0,0 +1,2562 @@
...
+  do
+    {
+      struct usb_cdc_ncm_ndp16_s *ndp
+        = (struct usb_cdc_ncm_ndp16_s *)(priv->rxnetbuf + ndpoffset);
+
+      ndpoffset = usbhost_getle16(ndp->next_ndp_index);
+
+      /* Parse each DPE */
+
+      struct usb_cdc_ncm_dpe16_s *dpe;
```
Comment (patacongo) [inline drivers/usbhost/usbhost_cdcmbim.c:866]: C89 violation.  Move to top of the compound statement.

File: drivers/usbhost/usbhost_cdcmbim.c (hunk 10 above)
Comment (acassis) [reply drivers/usbhost/usbhost_cdcmbim.c:365]: no, it has 78 chars

File: drivers/usbhost/usbhost_cdcmbim.c (hunk 11 above)
Comment (acassis) [reply drivers/usbhost/usbhost_cdcmbim.c:511]: 75

**Comment by patacongo** [comment]
> The driver itself is mostly useless without a corresponding user app to send the MBIM commands. However writing this type of program is very simple and the MBIM specification is fully open.

Do you have a sample program for testing purposes?

File: net/Kconfig [hunk 18]
```diff
@@ -150,6 +150,11 @@ config NET_LOOPBACK_PKTSIZE
 		CONFIG_NET_LOOPBACK_PKTSIZE is zero, meaning that this maximum
 		packet size will be used by loopback driver.
 
+menuconfig NET_MBIM
+        bool "MBIM modem support"
+	depends on USBHOST_CDCMBIM
+	default n
+
```
Comment (patacongo) [inline net/Kconfig:157]: Something wrong here.  Looks like TABs used on some lines and spaces on others?

Comment (acassis) [reply net/Kconfig:157]: Yes, it was with spaces in the first line, fixed!

**Comment by aport** [comment]
> @aport Do you have a sample program for testing purposes?

Unfortunately I don't have anything suitable for contribution at the moment. The software I was using to communicate with the modem was proprietary and ported from a Linux platform. I can ask about splitting the MBIM-specific code out and licensing Apache. Should have no issue with that since we decided to use QMI instead of MBIM (of course after I spent all the time writing the MBIM code)

And as you have noticed there are *a lot* of coding style violations in the original patch. I apologize for the poor quality but it was essentially a proof of concept that was later abandoned, and I posted it to the mailing list in hopes someone else found it useful to them.

**Comment by acassis** [comment]
> > @aport Do you have a sample program for testing purposes?
> 
> Unfortunately I don't have anything suitable for contribution at the moment. The software I was using to communicate with the modem was proprietary and ported from a Linux platform. I can ask about splitting the MBIM-specific code out and licensing Apache. Should have no issue with that since we decided to use QMI instead of MBIM (of course after I spent all the time writing the MBIM code)
> 
> And as you have noticed there are _a lot_ of coding style violations in the original patch. I apologize for the poor quality but it was essentially a proof of concept that was later abandoned, and I posted it to the mailing list in hopes someone else found it useful to them.

Adam, don't worry, you mostly followed the coding style. I didn't spend much time fixing it. I didn't know about these pitfalls before starting porting it. I hope we eventually get it working on mainline.

**Comment by patacongo** [comment]
> Do you have a sample program for testing purposes?

@aport No problem.  I am think of buying some hardware.  It there a low-cost CDC-MBIM USB device in a price range for an open source guy?  I am completely ignorant of the protocols, interfaces, etc but perhaps I will get sufficiently motivated.

**Comment by aport** [comment]
The easiest thing to do is probably pick up a mini-pcie to USB board similar to this one https://www.amazon.com/Adapter-Express-Converter-Support-Wireless/dp/B07FJXNLLT/ref=sr_1_1?dchild=1&keywords=mpcie+to+usb&qid=1592183911&sr=8-1

Then you can find a cheap module that supports the bands you need in your location. Almost every LTE module will support MBIM. From what I have seen, most mini-pcie modules will be in QMI mode out of the box and need to be switched into MBIM mode using a vendor-specific AT command. The setting is stored in nvram so you just need to do it once.

Every m.2 LTE card I have seen is in MBIM mode out of the box.

Quectel devices seem to be affordable and available on ebay etc. This code was tested using a Quectel EP06.

Also one final note, if you are testing this code using the STM32 FS or HS host controller (or one with the same IP), I think there is still one remaining bug in the host driver that will cause problems for this kind of device. The bug is related to the interpretation of a NAK as a zero-length packet when the amount of data transferred is a multiple of the endpoint size. I have a patch but again, the code is probably not correct. I will pass it to Alan for review.

I would recommend testing this driver on a target with a real EHCI controller first.

**Comment by btashton** [comment]
@patacongo as @aport said most 4G modems are a safe bet, the exception would be Ublox which I know the Toby line does not support it. 

I have some Sierra Wireless 7455 modems that are unlocked around from some previous work if you want one. The ZTE ones seem to have disappeared into a box, which is unfortunate since I found them much easier to work with. I never used MBIM mode when I was working with these.

**Comment by btashton** [comment]
Also that STM32 NAK issue is terrible, I never found a good way around it with the USB IP they use. I think I just ended up doing some ugly polling hack and moved to a different processor in the next hardware.  This was all 5 or so years ago so the details are fuzzy.

I would never recommend using the STM32 USB host. So many limitations and undocumented register bits ("documented" in the SDK and forums) that you have to set or the hardware does not even turn on.

**Pull request merged**